packages/ai-constructs/src/conversation/runtime/conversation_message_history_retriever.ts

```diff
@@ -0,0 +1,239 @@
+                          bytes
+                        }
+                      }
+                      image {
+                        format
+                        source {
+                          bytes
+                        }
+                      }
+                      json
+                      text
+                    }
+                    status
+                    toolUseId
+                  }
+                  toolUse {
+                    input
+                    name
+                    toolUseId
+                  }
+                }
+`;
+
+/**
+ * This class is responsible for retrieving message history that belongs to conversation turn event.
+ * It queries AppSync to list messages that belong to conversation.
+ * Additionally, it looks up a current message in case it's missing in the list due to eventual consistency.
+ */
+export class ConversationMessageHistoryRetriever {
+  /**
+   * Creates conversation message history retriever.
+   */
+  constructor(
+    private readonly event: ConversationTurnEvent,
+    private readonly graphqlRequestExecutor = new GraphqlRequestExecutor(
+      event.graphqlApiEndpoint,
+      event.request.headers.authorization
+    )
+  ) {}
+
+  getMessageHistory = async (): Promise<Array<ConversationMessage>> => {
+    if (this.event.messages?.length) {
+      // This is for backwards compatibility and should be removed with messages property.
+      return this.event.messages;
+    }
+    const messages = await this.listMessages();
+
+    let currentMessage = messages.find(
+      (m) => m.id === this.event.currentMessageId
+    );
+
+    // This is a fallback in case current message is not available in the message list.
+    // I.e. in a situation when freshly written message is not yet visible in
+    // eventually consistent reads.
+    if (!currentMessage) {
+      currentMessage = await this.getCurrentMessage();
+      messages.push(currentMessage);
+    }
+
+    // Index assistant messages by corresponding user message.
+    const assistantMessageByUserMessageId: Map<
+      string,
+      ConversationHistoryMessageItem
+    > = new Map();
+    messages.forEach((message) => {
+      if (message.role === 'assistant' && message.associatedUserMessageId) {
+        assistantMessageByUserMessageId.set(
+          message.associatedUserMessageId,
+          message
+        );
+      }
+    });
+
+    // Reconcile history and inject aiContext
+    return messages.reduce((acc, current) => {
+      // Bedrock expects that message history is user->assistant->user->assistant->... and so on.
+      // The chronological order doesn't assure this ordering if there were any concurrent messages sent.
+      // Therefore, conversation is ordered by user's messages only and corresponding assistant messages are inserted
+      // into right place regardless of their createdAt value.
+      // This algorithm assumes that GQL query returns messages sorted by createdAt.
```

Comment:
It's apparently not easy with NoSQL like DDB. It would require custom resolver to pretty much encapsulate same logic. @atierian is that correct?

Some other considerations:
1. Number of messages per conversatonId is expected to be rather small set (which we cap at 1000 items, might be configurable in the future).
2. We have to have all these messages in memory anyways to send it to Bedrock.
3. The DDB and query provide linear view if history audit by createdAt. The special ordering and filtering is for Bedrock API requirements. Appsync doesn't know about Bedrock calls, Lambda knows, so this adaptation of data seem to belong to Lambda.
4. This isn't one-way door. I.e. we can move ordering back to special list query and it's resolver later and take this out. Without breaking (multiple application of this algorithm is idempotent).

Reply:
It's possible to do this logic in the AppSync resolver function, however the downsides outweigh the benefits.

For context, there's a single GraphQL query per conversation route for listing messages (`listConversationMessage<RouteName>`). The list query is used by clients and now the Lambda function.

The ordering and omitting in-flight user messages added here is specific to Bedrock's API requirements, not intrinsic to querying message history. We don't want to omit in-flight / currently unanswered user messages from responses to that list query because it could lead to confusing behavior in the client application.

I see a couple options:
1. Bedrock specific history logic lives in lambda (these changes).
2. We introduce a flag in the list query for bimodal behavior. 
3. We introduce a separate GraphQL query for outputting messages in this structure.

Of those, 1. feels like the right choice for separation of responsibilities and maintenance.